src/normalization.md

```diff
@@ -0,0 +1,314 @@
+
+We consider types to be rigid if their "shape" isn't going to change, for example `Box` is rigid as no amount of normalization can turn a `Box` into a `u32`, whereas `<vec::IntoIter<u32> as Iterator>::Item` is not rigid as it can be normalized to `u32`.
+
+Aliases are rigid when we know that we will never be able to normalize them further. A concrete example of a *rigid* alias would be `<T as Iterator>::Item` in an environment where there is no `T: Iterator<Item = ...>` bound, only a `T: Iterator` bound:
+```rust
+fn foo<T: Iterator>() {
+    // This alias is *rigid*
+    let _: <T as Iterator>::Item;
+}
+
+fn bar<T: Iterator<Item = u32>>() {
+    // This alias is *not* rigid as it can be normalized to `u32`
+    let _: <T as Iterator>::Item;
+}
+```
+
+This definition also implies that illformed aliases can be considered rigid. For example the type `<u32 as Iterator>::Item` is a rigid alias as it can never be normalized as there is no `u32: Iterator` impl (ignoring `feature(trivial_bounds)`).
+
+When an alias can't yet be normalized but may eventually wind up normalizable we consider it to be an "ambiguous" alias. This can occur when an alias contains inference variables which prevent being able to determine how the trait is implemented:
+```rust
+fn foo<T: Iterator, U: Iterator>() {
+    // This alias is considered to be "ambiguous"
+    let _: <_ as Iterator>::Item;
+}
+```
+
+The reason we call them "ambiguous" aliases is because its *ambiguous* whether this is a rigid alias or not.
+
+The source of the `_: Iterator` trait impl is *ambiguous* (i.e. unknown), it could be some `impl Iterator for u32` or it could be some `T: Iterator` trait bound, we don't know yet. Depending on how `_: Iterator` holds the alias could be an unnormalized alias or it could be a rigid alias; it's *ambiguous* what kind of alias this is.
+
+Finally, an alias can just be unnormalized, `<Vec<u32> as IntoIterator>::Iter` is an unnormalized alias as it can already be normalized to `std::vec::IntoIter<u32>`, it just hasn't been done yet.
+
+---
+
+It is worth noting that Free and Inherent aliases cannot be rigid or ambiguous as   naming them also implies having resolved the definition of the alias, which specifies the underlying type of the alias:
+```rust
+#![feature(inherent_associated_types)]
+
+trait Trait {}
+
+impl Trait for u32 {}
+impl Trait for u64 {}
+
+struct Foo<T>(T);
+
+impl<T: Trait> Foo<T> {
+    type Inherent = u32;
+}
+
+fn ret_alias<T: Trait>(_: T) -> Foo<T>::Inherent {
+    10_u32
+}
+
+fn main() {
+    // `t` has type `?x`
+    let mut t = Default::default();
+    // `ret_alias` returns a value of type `Foo<?x>::Inherent`
+    t = ret_alias(t);
+    
+    // we have equated `?x` with `Foo<?x>::Inherent` this can
+    // only succeed if we're able to normalize inherent aliases
+    // with ambiguous where clauses on the impl.
+    //
+    // Or, put another way, it is *not* an ambiguous alias.
+}
+```
```

Comment:
What are you trying to show with this example? It feels confusing to me (and too "large" for showing that inherent aliases are always normalizeable)

Reply:
Yeah idk removed it